packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -4493,3 +4523,57 @@ export class ContainerRuntime
 		return this.sessionSchema.opGroupingEnabled === true;
 	}
 }
+
+/**
+ * Represents the runtime of the container. Contains helper functions/state of the container.
+ * It will define the store level mappings.
+ */
+class ContainerRuntimeInternal extends ContainerRuntime implements IRuntimeInternal {
```

Comment:
Please avoid class inheritance as the mechanism to add routes - at minimum this is challenging to scale to further extended functionality along with the usual downsides of class inheritance.

Reply:
Open to alternative. ContainerRuntime is `@legacy` and none of the rest of this should be exposed (my current opinion).

Reply:
The only way for container runtime author to create a container is by using a concrete implementation of ContainerRuntime today. We can change that and expose a factory, and such factory could take a bunch of args to decide what functionality to mix in, but this does not exist today. Plus, I doubt that IContainerRuntime has all the right things exposed on it to be a viable replacement for ContainerRuntime type.

You can easily see the size of the problem by looking at Bohemia repo.

In other places we solved that same problem via mixins (i.e. dynamically created class types), allow arbitrary inheritance. It solves some problems, leaves other problems not solved. I personally prefer mixin approach overexposing 100 hooks, but I also see the opposite point of view (harder to see contracts, etc.).

Reply:
I admit to not understanding what is really supported with `ContainerRuntime` today and what we want.
As far as I can tell no one should create `ContainerRuntime` directly. They should be passing factory (`containerRuntimeCtor`) to `ContainerRuntime.loadRuntime`, no? I see 0 [hits for `containerRuntimeCtor` in Bohemia](https://office.visualstudio.com/OC/_search?action=contents&text=containerRuntimeCtor&type=code). Mixins are okay, but I still don't know what we want to expose or not. I really don't have a picture of where we are going; so, I can't fully judge if I am recommending an appropriate route.
While new feature I need support for remains experimental, I think it is okay to not have the full internal structure settled. It should be okay for someone doing something custom with `ContainerRuntime` to not get this support because is not settled.
All that said I switched to a mixin pattern here.